op-e2e/actions/interop/interop_test.go

```diff
@@ -467,3 +467,92 @@ func TestInteropExecutingMessageOutOfRangeLogIndex(gt *testing.T) {
 	assertHeads(t, actors.ChainA, 1, 0, 0, 0)
 	assertHeads(t, actors.ChainB, 1, 0, 1, 0)
 }
+
+func TestInterop_IntraBlockReferenceReplacement(gt *testing.T) {
+	t := helpers.NewDefaultTesting(gt)
+	system := dsl.NewInteropDSL(t)
+	actors := system.Actors
+	alice := system.CreateUser()
+	emitterContract := dsl.NewEmitterContract(t)
+	system.AddL2Block(actors.ChainA, dsl.WithL2BlockTransactions(
+		emitterContract.Deploy(alice),
+	))
+	system.AddL2Block(actors.ChainB, dsl.WithL2BlockTransactions(
+		emitterContract.Deploy(alice),
+	))
+	assertHeads(t, actors.ChainA, 1, 0, 1, 0)
+	assertHeads(t, actors.ChainB, 1, 0, 1, 0)
+
+	// Create a scenario where ChainB has a block with an invalid message and ChainA references it.
+	// The blocks on both chains should be marked as invalid and replaced.
+	var (
+		chainAExecTx *dsl.GeneratedTransaction
+		chainBExecTx *dsl.GeneratedTransaction
+		chainBInitTx *dsl.GeneratedTransaction
+	)
+	{
+		actors.ChainA.Sequencer.ActL2StartBlock(t)
+		actors.ChainB.Sequencer.ActL2StartBlock(t)
+
+		chainAInitTx := emitterContract.EmitMessage(alice, "chainA message")(actors.ChainA)
+		chainAInitTx.Include()
+
+		// Create messages with a conflicting payload on chain B, while also emitting an initiating message
+		chainBExecTx = system.InboxContract.Execute(alice, chainAInitTx,
+			dsl.WithPayload([]byte("this message was never emitted")))(actors.ChainB)
+		chainBExecTx.Include()
+		chainBInitTx = emitterContract.EmitMessage(alice, "chainB message")(actors.ChainB)
+		chainBInitTx.Include()
+
+		// Create a message with a valid message on chain A, pointing to the initiating message on B from the same block
+		// as an invalid message.
+		chainAExecTx = system.InboxContract.Execute(alice, chainBInitTx)(actors.ChainA)
+		chainAExecTx.Include()
+
+		actors.ChainA.Sequencer.ActL2EndBlock(t)
+		actors.ChainB.Sequencer.ActL2EndBlock(t)
+		assertHeads(t, actors.ChainA, 2, 0, 1, 0)
+		assertHeads(t, actors.ChainB, 2, 0, 1, 0)
+	}
+
+	// Submit the data to the L1
+	// Now the local-safe head of both chains is 1 but the cross-safe head is still 0
+	system.SubmitBatchData(func(opts *dsl.SubmitBatchDataOpts) {
+		opts.SkipCrossSafeUpdate = true
+	})
+	assertHeads(t, actors.ChainA, 2, 2, 1, 0)
+	assertHeads(t, actors.ChainB, 2, 2, 1, 0)
+
+	// Get the current local-safe blocks on both chains; these should be the invalid blocks
+	statusA, statusB := actors.ChainA.Sequencer.SyncStatus(), actors.ChainB.Sequencer.SyncStatus()
+	initialChainABlock := statusA.LocalSafeL2
+	initialChainBBlock := statusB.LocalSafeL2
+
+	// Sync the supervisor and process the full state
+	actors.ChainA.Sequencer.SyncSupervisor(t)
+	actors.ChainB.Sequencer.SyncSupervisor(t)
+	actors.Supervisor.ProcessFull(t)
+	actors.ChainA.Sequencer.ActL2PipelineFull(t)
+	actors.ChainA.Sequencer.SyncSupervisor(t)
```

Comment:
I see that this is the missing call that fixes the test. The supervisor needs to be synchronized between each chain's derivation processing.
So the test had violated an implicit ordering of events that the supervisor expected in this scenario. This invalid sequence of events probably only worked by coincidence in the happy path case.

Reply:
The happy path likely succeeded for real since it didn't require a replacement block. The key here is that handling a block replacement is a 3-phase exchange:
1. Invalidation/Replacement is detected by the Supervisor -> Instruct the Node to Reset
1. Node Invalidates the Block -> Inform the Supervisor of the new block
1. Supervisor takes the new replacement and resolves the database with it

While steps 1 and 3 can be handled by `ProcessFull`, which drains the event-queue, step 2 is an RPC call that needs to travel from the Node to the Supervisor. That's exactly what `SyncSupervisor` does, it takes all the pending RPC messages from the Node, and loads them to the Supervisor.